File: license-bot/bot.py
import os

import discord
from discord import app_commands

import db

ADMIN_USER_IDS = {
    int(x) for x in os.environ.get("ADMIN_USER_IDS", "").split(",") if x.strip()
}
ADMIN_ROLE_ID = os.environ.get("ADMIN_ROLE_ID")
ADMIN_ROLE_ID = int(ADMIN_ROLE_ID) if ADMIN_ROLE_ID else None

intents = discord.Intents.default()
client = discord.Client(intents=intents)
tree = app_commands.CommandTree(client)


def is_admin(interaction: discord.Interaction) -> bool:
    if interaction.user.id in ADMIN_USER_IDS:
        return True
    if ADMIN_ROLE_ID and isinstance(interaction.user, discord.Member):
        return any(role.id == ADMIN_ROLE_ID for role in interaction.user.roles)
    return False


def _require_admin(interaction: discord.Interaction) -> bool:
    return is_admin(interaction)


def _status_line(lic) -> str:
    if lic["revoked"]:
        status = "REVOKED"
    elif lic["bound_uuid"]:
        status = f"bound to `{lic['bound_username'] or lic['bound_uuid']}`"
    else:
        status = "unbound (not yet activated)"
    return f"`{lic['key']}` — {status}"


@tree.command(description="[admin] Generate a new license key for a member")
@app_commands.describe(member="Who this key is for", note="Optional note (e.g. order id)")
async def genkey(interaction: discord.Interaction, member: discord.Member, note: str | None = None):
    if not _require_admin(interaction):
        await interaction.response.send_message("You can't use this command.", ephemeral=True)
        return

    key = db.create_license(str(member.id), note)

    dm_sent = True
    try:
        await member.send(
            f"You've been issued a TPA Tools license key:\n`{key}`\n\n"
            "Put this in your mod config. It activates on the first Minecraft "
            "account that uses it and won't work on any other account after that, "
            "so don't share it."
        )
    except discord.Forbidden:
        dm_sent = False

    msg = f"Created key `{key}` for {member.mention}."
    if not dm_sent:
        msg += "\n⚠️ Couldn't DM them (DMs closed) — you'll need to send it manually."
    await interaction.response.send_message(msg, ephemeral=True)


@tree.command(description="See your own license keys and their status")
async def mykeys(interaction: discord.Interaction):
    licenses = db.licenses_for_user(str(interaction.user.id))
    if not licenses:
        await interaction.response.send_message("You have no license keys.", ephemeral=True)
        return
    lines = [_status_line(lic) for lic in licenses]
    await interaction.response.send_message("\n".join(lines), ephemeral=True)


@tree.command(description="[admin] Look up a license key")
@app_commands.describe(key="The license key")
async def lookup(interaction: discord.Interaction, key: str):
    if not _require_admin(interaction):
        await interaction.response.send_message("You can't use this command.", ephemeral=True)
        return
    lic = db.get_license(key)
    if lic is None:
        await interaction.response.send_message("No such key.", ephemeral=True)
        return
    await interaction.response.send_message(
        f"Key: `{lic['key']}`\n"
        f"Owner: <@{lic['discord_user_id']}>\n"
        f"Status: {_status_line(lic)}\n"
        f"Mismatch attempts: {lic['mismatch_attempts']}\n"
        f"Note: {lic['note'] or '-'}",
        ephemeral=True,
    )


@tree.command(description="[admin] Revoke a license key")
@app_commands.describe(key="The license key to revoke")
async def revoke(interaction: discord.Interaction, key: str):
    if not _require_admin(interaction):
        await interaction.response.send_message("You can't use this command.", ephemeral=True)
        return
    ok = db.revoke_license(key)
    await interaction.response.send_message(
        "Revoked." if ok else "No such key.", ephemeral=True
    )


@tree.command(description="[admin] Unbind a key so it can activate on a new Minecraft account")
@app_commands.describe(key="The license key to unbind")
async def unbind(interaction: discord.Interaction, key: str):
    if not _require_admin(interaction):
        await interaction.response.send_message("You can't use this command.", ephemeral=True)
        return
    ok = db.unbind_license(key)
    await interaction.response.send_message(
        "Unbound — it'll activate on whichever account uses it next." if ok else "No such key.",
        ephemeral=True,
    )


@client.event
async def on_ready():
    guild_id = os.environ.get("DISCORD_GUILD_ID")
    if guild_id:
        guild = discord.Object(id=int(guild_id))
        tree.copy_global_to(guild=guild)
        await tree.sync(guild=guild)
    else:
        await tree.sync()
    print(f"Logged in as {client.user}")

File: license-bot-node/bot.js
const {
  Client,
  GatewayIntentBits,
  SlashCommandBuilder,
  REST,
  Routes,
  PermissionFlagsBits,
} = require('discord.js');
const db = require('./db');

const ADMIN_USER_IDS = new Set(
  (process.env.ADMIN_USER_IDS || '')
    .split(',')
    .map((s) => s.trim())
    .filter(Boolean)
);
const ADMIN_ROLE_ID = process.env.ADMIN_ROLE_ID || null;

// Off by default: treat anyone with the server's own Administrator
// permission as a bot admin too, on top of ADMIN_USER_IDS/ADMIN_ROLE_ID.
// Only turn this on for servers where you trust everyone with that
// permission to hand out and revoke license keys.
const TRUST_SERVER_ADMINS = process.env.TRUST_SERVER_ADMINS === 'true';

const client = new Client({ intents: [GatewayIntentBits.Guilds] });

function isAdmin(interaction) {
  if (ADMIN_USER_IDS.has(interaction.user.id)) return true;
  if (ADMIN_ROLE_ID && interaction.member?.roles?.cache?.has(ADMIN_ROLE_ID)) return true;
  if (TRUST_SERVER_ADMINS && interaction.memberPermissions?.has(PermissionFlagsBits.Administrator)) {
    return true;
  }
  return false;
}

function statusLine(lic) {
  let status;
  if (lic.revoked) status = 'REVOKED';
  else if (lic.bound_uuid) status = `bound to \`${lic.bound_username || lic.bound_uuid}\``;
  else status = 'unbound (not yet activated)';
  return `\`${lic.key}\` — ${status}`;
}

const commands = [
  new SlashCommandBuilder()
    .setName('genkey')
    .setDescription('[admin] Generate a new license key for a member')
    .addUserOption((o) => o.setName('member').setDescription('Who this key is for').setRequired(true))
    .addStringOption((o) => o.setName('note').setDescription('Optional note (e.g. order id)')),
  new SlashCommandBuilder().setName('mykeys').setDescription('See your own license keys and their status'),
  new SlashCommandBuilder()
    .setName('lookup')
    .setDescription('[admin] Look up a license key')
    .addStringOption((o) => o.setName('key').setDescription('The license key').setRequired(true)),
  new SlashCommandBuilder()
    .setName('revoke')
    .setDescription('[admin] Revoke a license key')
    .addStringOption((o) => o.setName('key').setDescription('The license key to revoke').setRequired(true)),
  new SlashCommandBuilder()
    .setName('unbind')
    .setDescription("[admin] Unbind a key so it can activate on a new Minecraft account")
    .addStringOption((o) => o.setName('key').setDescription('The license key to unbind').setRequired(true)),
  new SlashCommandBuilder()
    .setName('addhash')
    .setDescription('[admin] Register a known-good jar hash for the build you just shipped')
    .addStringOption((o) => o.setName('sha256').setDescription('SHA-256 of the jar').setRequired(true))
    .addStringOption((o) => o.setName('label').setDescription('e.g. version number')),
  new SlashCommandBuilder()
    .setName('removehash')
    .setDescription('[admin] Remove a known-good jar hash')
    .addStringOption((o) => o.setName('sha256').setDescription('SHA-256 to remove').setRequired(true)),
  new SlashCommandBuilder().setName('hashes').setDescription('[admin] List registered known-good jar hashes'),
  new SlashCommandBuilder()
    .setName('tamperlog')
    .setDescription('[admin] Show recent tampered-jar detections')
    .addIntegerOption((o) => o.setName('limit').setDescription('How many to show (default 10)')),
].map((c) => c.toJSON());

async function registerCommands() {
  const rest = new REST().setToken(process.env.DISCORD_BOT_TOKEN);
  const appId = client.application.id;
  if (process.env.DISCORD_GUILD_ID) {
    await rest.put(Routes.applicationGuildCommands(appId, process.env.DISCORD_GUILD_ID), {
      body: commands,
    });
  } else {
    await rest.put(Routes.applicationCommands(appId), { body: commands });
  }
}

client.once('ready', async () => {
  await registerCommands();
  console.log(`Logged in as ${client.user.tag}`);
});

client.on('interactionCreate', async (interaction) => {
  if (!interaction.isChatInputCommand()) return;

  try {
    switch (interaction.commandName) {
      case 'genkey': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const member = interaction.options.getUser('member', true);
        const note = interaction.options.getString('note');
        const key = db.createLicense(member.id, note);

        let dmSent = true;
        try {
          await member.send(
            `You've been issued a TPA Tools license key:\n\`${key}\`\n\n` +
              "Put this in your mod config. It activates on the first Minecraft " +
              "account that uses it and won't work on any other account after that, " +
              "so don't share it."
          );
        } catch {
          dmSent = false;
        }

        let msg = `Created key \`${key}\` for <@${member.id}>.`;
        if (!dmSent) msg += "\n⚠️ Couldn't DM them (DMs closed) — you'll need to send it manually.";
        return interaction.reply({ content: msg, ephemeral: true });
      }

      case 'mykeys': {
        const licenses = db.licensesForUser(interaction.user.id);
        if (!licenses.length) {
          return interaction.reply({ content: 'You have no license keys.', ephemeral: true });
        }
        return interaction.reply({ content: licenses.map(statusLine).join('\n'), ephemeral: true });
      }

      case 'lookup': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const key = interaction.options.getString('key', true);
        const lic = db.getLicense(key);
        if (!lic) return interaction.reply({ content: 'No such key.', ephemeral: true });
        return interaction.reply({
          content:
            `Key: \`${lic.key}\`\n` +
            `Owner: <@${lic.discord_user_id}>\n` +
            `Status: ${statusLine(lic)}\n` +
            `Mismatch attempts: ${lic.mismatch_attempts}\n` +
            `Note: ${lic.note || '-'}`,
          ephemeral: true,
        });
      }

      case 'revoke': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const key = interaction.options.getString('key', true);
        const ok = db.revokeLicense(key);
        return interaction.reply({ content: ok ? 'Revoked.' : 'No such key.', ephemeral: true });
      }

      case 'unbind': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const key = interaction.options.getString('key', true);
        const ok = db.unbindLicense(key);
        return interaction.reply({
          content: ok ? "Unbound — it'll activate on whichever account uses it next." : 'No such key.',
          ephemeral: true,
        });
      }

      case 'addhash': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const sha256 = interaction.options.getString('sha256', true).trim();
        if (!/^[0-9a-fA-F]{64}$/.test(sha256)) {
          return interaction.reply({ content: 'That doesn\'t look like a SHA-256 hash (64 hex chars).', ephemeral: true });
        }
        const label = interaction.options.getString('label');
        db.addKnownHash(sha256, label);
        return interaction.reply({ content: `Registered \`${sha256}\`${label ? ` (${label})` : ''} as known-good.`, ephemeral: true });
      }

      case 'removehash': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const sha256 = interaction.options.getString('sha256', true).trim();
        const ok = db.removeKnownHash(sha256);
        return interaction.reply({ content: ok ? 'Removed.' : 'No such hash.', ephemeral: true });
      }

      case 'hashes': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const rows = db.listKnownHashes();
        if (!rows.length) {
          return interaction.reply({
            content: 'No known-good hashes registered yet — every jar hash will be treated as untampered until you add one.',
            ephemeral: true,
          });
        }
        return interaction.reply({
          content: rows.map((r) => `\`${r.sha256}\`${r.label ? ` — ${r.label}` : ''}`).join('\n'),
          ephemeral: true,
        });
      }

      case 'tamperlog': {
        if (!isAdmin(interaction)) {
          return interaction.reply({ content: "You can't use this command.", ephemeral: true });
        }
        const limit = interaction.options.getInteger('limit') ?? 10;
        const rows = db.recentTamperLog(limit);
        if (!rows.length) {
          return interaction.reply({ content: 'No tamper detections logged.', ephemeral: true });
        }
        const lines = rows.map(
          (r) =>
            `<t:${Math.floor(r.detected_at)}:R> key \`${r.key || '?'}\` uuid \`${r.minecraft_uuid || '?'}\` hash \`${r.jar_sha256}\``
        );
        return interaction.reply({ content: lines.join('\n'), ephemeral: true });
      }
    }
  } catch (err) {
    console.error(err);
    if (interaction.isRepliable()) {
      await interaction.reply({ content: 'Something went wrong.', ephemeral: true }).catch(() => {});
    }
  }
});

async function notifyTamper({ key, uuid, username, jarHash }) {
  const message =
    `⚠️ **Tampered jar detected**\n` +
    `Key: \`${key || 'unknown'}\`\n` +
    `Minecraft account: \`${username || uuid}\`\n` +
    `Jar SHA-256: \`${jarHash}\` (not a registered known-good build)`;

  for (const adminId of ADMIN_USER_IDS) {
    try {
      const user = await client.users.fetch(adminId);
      await user.send(message);
    } catch (e) {
      console.error(`Could not DM admin ${adminId} about tamper detection:`, e.message);
    }
  }
}

module.exports = { client, notifyTamper };
